pylanetary/spectral/core.py

```diff
@@ -0,0 +1,808 @@
+import matplotlib.pyplot as plt
+from matplotlib.patches import Ellipse,Circle
+from matplotlib import colors
+from matplotlib.colors import LinearSegmentedColormap
+import matplotlib
+import scipy.io
+from astropy.io import fits
+from astropy.stats import sigma_clipped_stats
+from astropy.constants import c
+from pylanetary.utils import *
+import time
+import pdb
+
+def centered_list(n):
+    if n % 2 == 0: # If n is even, we adjust it to the next odd number
+        n += 1
+    half = n // 2
+    centered_list = list(range(-half, half + 1,2))
+    return centered_list
+
+def calc_doppler_vel(peak, err, rest_f):
+    '''
+    Calculate the Doppler velocity using the Doppler Equation.
+    
+    Parameters
+    ---------- 
+    peak : float
+          The calculated central peak frequency of the spectral line peak.
+    err : float
+          The standard deviation of the calculated central peak frequency.
+    rest_f : float
+          The rest frequency of the transition.
+    
+    Returns
+    -------
+    v : float
+          The calculated Doppler velocity in m/s.
+    wind_err : float
+          The Doppler velocity error, calculated through error propagation, in m/s.
+    
+    '''
+    a = c.value/rest_f
+    
+    v = ((-1*peak/rest_f)+1)*c.value
+    print('The wind speed is '+str(v)+'m/s')
+    wind_err = np.abs(a*err)
+    print('The wind error is '+str(wind_err)+'m/s')
+    
+    return v, wind_err
+    
+model = {}
+model['Gaussian'] = {'ref': 'gauss', 'lmfit_func': GaussianModel(),'label': 'Gaussian Fit','color':'orange'}
+model['Lorentzian'] = {'ref': 'lor','lmfit_func':LorentzianModel(),'label': 'Lorentzian Fit','color':'green'}
+model['Voigt'] = {'ref': 'voigt','lmfit_func':VoigtModel(),'label': 'Voigt Fit','color':'red'}
+model['PseudoV'] = {'ref': 'psv','lmfit_func':PseudoVoigtModel(),'label': 'Pseudo-Voigt Fit','color':'purple'}
+model['Moffat'] = {'ref': 'mof','lmfit_func':MoffatModel(),'label': 'Moffat Fit','color':'brown'}
+
+###############################################################################
+###############################################################################
+
+class Spectrum:
+  '''
+  Analysis tools for a single spectrum.
+    
+  Example usage:
+  
+  from pylanetary.spectral import Spectrum
+  your_spec = Spectrum(spec, freq = freq,rest_f = 345.79598990,RMS = 0.017530593)
+  
+  test_peak, test_std, chi, chi_red = test.fit_profile(fit_type = 'Moffat', plot_name = 'line_comparison.png')
+  
+  test_wind, test_err = test.calc_doppler_vel(peak = test_peak, err = test_std)
+  '''
+  
+  ###############################################################################
+  
+  def __init__(self, spec, rest_f, x_space, wl = None, freq = None, RMS = None):
+    '''
+    Input string instantiates object using numpy arrays of the x-axis and the spectral intensity axis.
+    
+    Parameters
+    ----------
+    spec : numpy array
+          The intensity/vertical axis of the spectrum
+    rest_f : float
```

Comment:
I don't think the rest frequency needs to be in the __init__ of Spectrum; for example, what if you wanted to fit two molecular lines in the same spectrum, or you wanted to test the effect of perturbing the rest frequency a little bit?  This could instead be an input parameter to fit_profile.